src/tools/resourceAccessChecker.ts

```diff
@@ -0,0 +1,365 @@
+          },
+        });
+
+        if (!this.allowedProjectIds.has(datasourceProjectId)) {
+          return {
+            allowed: false,
+            message: [
+              'The set of allowed data sources that can be queried is limited by the server configuration.',
+              `The datasource with LUID ${datasourceLuid} cannot be queried because it does not belong to an allowed project.`,
+            ].join(' '),
+          };
+        }
+      } catch (error) {
+        return {
+          allowed: false,
+          message: [
+            'The set of allowed data sources that can be queried is limited by the server configuration.',
+            `An error occurred while checking if the datasource with LUID ${datasourceLuid} is in an allowed project:`,
+            getExceptionMessage(error),
+          ].join(' '),
+        };
+      }
+    }
+
+    return { allowed: true };
+  }
+
+  private async _isWorkbookAllowed({
+    workbookId,
+    restApiArgs: { config, requestId, server },
+  }: {
+    workbookId: string;
+    restApiArgs: RestApiArgs;
+  }): Promise<AllowedResult<Workbook>> {
+    const cachedResult = this._cachedWorkbookIds.get(workbookId);
+    if (cachedResult) {
+      return cachedResult;
+    }
+
+    if (this.allowedWorkbookIds && !this.allowedWorkbookIds.has(workbookId)) {
+      return {
+        allowed: false,
+        message: [
+          'The set of allowed workbooks that can be queried is limited by the server configuration.',
+          `Querying the workbook with LUID ${workbookId} is not allowed.`,
+        ].join(' '),
+      };
+    }
+
+    let workbook: Workbook | undefined;
+    if (this.allowedProjectIds) {
+      try {
+        workbook = await useRestApi({
+          config,
+          requestId,
+          server,
+          jwtScopes: ['tableau:content:read'],
+          callback: async (restApi) => {
+            const workbook = await restApi.workbooksMethods.getWorkbook({
+              siteId: restApi.siteId,
+              workbookId,
+            });
+
+            return workbook;
+          },
+        });
+
+        if (!this.allowedProjectIds.has(workbook.project?.id ?? '')) {
+          return {
+            allowed: false,
+            message: [
+              'The set of allowed workbooks that can be queried is limited by the server configuration.',
+              `The workbook with LUID ${workbookId} cannot be queried because it does not belong to an allowed project.`,
+            ].join(' '),
+          };
+        }
+      } catch (error) {
+        return {
+          allowed: false,
+          message: [
+            'The set of allowed workbooks that can be queried is limited by the server configuration.',
+            `An error occurred while checking if the workbook with LUID ${workbookId} is in an allowed project:`,
+            getExceptionMessage(error),
+          ].join(' '),
+        };
+      }
+    }
+
+    return { allowed: true, content: workbook };
```

Comment:
Here's that "special" thing I mentioned in comment: https://github.com/tableau/tableau-mcp/pull/131/files#r2458160645

This case doesn't apply for the other content types which is why you don't see it for data sources or views.